Project template

barcharcraz opened the pull request:
This is a template for projects

said template is in the project_template directory and includes the project skeleton, snippets for copyright and license header (MIT using an SPDX identifier), as well as a basic example of documentation.

Note: there is no process to actually fetch libraries to embed since I didn't want to add a git submodule to the parent project.

also, please squash when merging.

barcharcraz commented:
@microsoft/vclibs

@ras0219-msft

CaseyCarter commented on .vscode/snippets.code-snippets:
```diff
@@ -0,0 +1,26 @@
+		]
+	}
+	
+}
```
```suggestion
}

```

CaseyCarter commented on .vscode/snippets.code-snippets:
```diff
@@ -0,0 +1,26 @@
+{
+	// Place your azure-sdk-for-c workspace snippets here. Each snippet is defined under a snippet name and has a scope, prefix, body and 
+	// description. Add comma separated ids of the languages where the snippet is applicable in the scope field. If scope 
+	// is left empty or omitted, the snippet gets applied to all languages. The prefix is what is 
```
Inconsistent line wrapping in the above.

CaseyCarter commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,12 @@
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
+- **extra-cmake-modules**: https://github.com/KDE/extra-cmake-modules.git
+
+these may be installed using a package manager or embedded in the external/ folder.
```
```suggestion
These may be installed using a package manager or embedded in the `external` folder.
```

CaseyCarter commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,12 @@
+
+## Requirements
+
+This azure sdk module has several external build-time requirements.
```
```suggestion
This Azure SDK module has several external build-time requirements:
```

CaseyCarter commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+
+# this is a cmake 3.15 feature that tells find_package to 
+# look at config files before module files. This makes it easier
+# to write buildsystems that work on many platforms, espicially
```
```suggestion
# to write buildsystems that work on many platforms, especially
```

CaseyCarter commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+
+
+
+# this is a cmake 3.15 feature that tells find_package to 
```
I can't tell what "this" is referring to here. (Is it the enormous block of empty lines above? There must be *some* reason for all those empty lines.)

CaseyCarter commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+set(CMAKE_CXX_STANDARD 17)
+
+# default to hidden symbols when building shared libraries
+# on unix systems
```
Does the below have different behavior on non-unix systems? If not, we should simply strike this line.

barcharcraz replied:
on windows, there is no way to build with "default visibility" as exported. This makes the unix behavior more consistent with the windows behavior (and more correct)

CaseyCarter replied:
So how about `# symbols in built shared libraries are hidden by default`?

CaseyCarter commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)
+
+set_property(DIRECTORY PROPERTY COMPILE_OPTIONS
+    $<$<COMPILE_LANG_AND_ID:C,MSVC>:/W4 /wd4201>
```
Notably `LANGUAGES` includes `CXX`, but we're only setting `COMPILE_OPTIONS` here for `C`. Should the two be consistent?

joshfree replied:
@CaseyCarter should this also be /Za and not /wd4201.

barcharcraz replied:
/wd4201 is (a superset of) a C11 feature, and is very useful and widely supported.

/Za really not a good idea, C99 (ish) compatibility should be enforced by compiling with warnings using compilers that actually support C99, which msvc does not. /Za also prevents you from including most windows system headers.

/W4 warns for all of the .... less than ideal ... ms extensions.

barcharcraz replied:
@CaseyCarter also: I'm not really sure how to handle this for c++, since technically you could compile with a c++ compiler that was different from your C compiler. But yes both should be set..

Also: the warnings for C files may well be different from the warnings for C++ files (tests).

CaseyCarter commented on project_template/.gitattributes:
```diff
@@ -0,0 +1 @@
+* text=auto eol=lf
```
\** Seething hate for git's EoL translation. **

ras0219-msft replied:
This should be `* -text`.

barcharcraz replied:
ugh, yeah. It's way to hard to tell git to just keep things the way they are.

CaseyCarter commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,12 @@
+
+these may be installed using a package manager or embedded in the external/ folder.
+
+By default dependencies will be found via the system, to enable embedding set
```
```suggestion
By default, dependencies will be found via the system; to enable embedding set
```

CaseyCarter commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+
+
+file(GLOB_RECURSE az_exampleshortname_src CONFIGURE_DEPENDS src/*.c)
+file(GLOB_RECURSE az_exampleshortname_test_src CONFIGURE_DEPENDS test/*.cpp)
```
I have a vague recollection that globs in the build system are to be avoided; they're only updated on actual cmake runs, right?

ras0219-msft replied:
I recommend globs.

While this means that yes, adding a file to your disk won't automatically update your build, not using globs won't automatically update your build either! If you use globs, then you just need to trigger a reconfigure through any means (simply running `cmake .` in the build directory or hitting `Ctrl-S` on the CMakeLists.txt in visual studio). If you use an explicit sources list, you must edit the build file AND trigger a reconfigure.

Globs are also highly beneficial if someone wants to consume the project from an alternative buildsystem (bazel, pants, boost build, make, scons, msbuild, etc). This is because globs ensure that your build is _simple_ -- `*.c`. If you use an explicit list, it is extremely easy to accidentally introduce unwanted complexity. This also reduces cognitive load on anyone who needs to read the build files because there is only one source of truth: the files on disk.

barcharcraz replied:
also, for most generators CONFIGURE_DEPENDS means adding a file will actually trigger a reconfigure of cmake.

CaseyCarter replied:
> If you use an explicit sources list, you must edit the build file AND trigger a reconfigure.

You lost me at "AND" here. Useful generators automatically reconfigure when the build is updated.

> If you use an explicit list, it is extremely easy to accidentally introduce unwanted complexity.

I suspect I'm biased by my experience which is heavily weighted toward header-only libraries in which every source file is a single-TU unit test.

> there is only one source of truth: the files on disk.

Which is all fine and dandy until you accidentally ship `tmp_hack.c`. ;)

In any case, I bow to the combination of Robert's expertise and Charlie's pointing out that my recollection predates `CONFIGURE_DEPENDS`. Feel free to resolve this comment.

barcharcraz replied:
yeah so what convinced me is that with file(GLOB_RECURSE) you ensure (by construction) that your build system works by doing essentially `gcc src/*.c` which is a _really_ nice property to have. And the difference even between listing sources and globbing without `CONFIGURE_DEPENDS` is just that you `touch` the buildsystem instead of finding where to list the file and listing it.

The buildsystem doesn't have to like, inotify on your source directory either (hi bezel 🙌!), it just runs `ls -f` and compares output.

CaseyCarter commented on project_template/LICENSE:
```diff
@@ -0,0 +1,21 @@
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE
```
```suggestion
    SOFTWARE.
```

joshfree replied:
good catch; other than the missing period, the text matched https://en.wikipedia.org/wiki/MIT_License.

barcharcraz replied:
I copy-pasted that from the LICENSE at the root of the repo.

CaseyCarter replied:
Fix that copy as well?

CaseyCarter commented on project_template/external/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+cmake_dependent_option(az_exampleshortname_embed_ecm "use embedded extra-cmake-modules" ${az_exampleshortname_embed_default} 
+    "az_exampleshortname_allow_embed" OFF)
+
+#this makes set() override option() even in projects
```
```suggestion
# this makes set() override option() even in projects
```

CaseyCarter commented on project_template/external/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+    set(CATCH_BUILD_TESTING OFF)
+    set(CATCH_INSTALL_DOCS OFF)
+    set(CATCH_INSTALL_HELPERS OFF)
+    # using exclude from all means doctest won't be
```
```suggestion
    # Using EXCLUDE_FROM_ALL means doctest won't be
```

CaseyCarter commented on project_template/external/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+    set(CATCH_INSTALL_HELPERS OFF)
+    # using exclude from all means doctest won't be
+    # built by default. However if a library from doctest is
+    # linked to then that library will be built.
```
Would "EXCLUDE_FROM_ALL means doctest won't be built unless some target links to a library from doctest." be clearer?

sergey-shandar commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Is there a way to share parts of this file across different projects? Something like `include` ?

barcharcraz replied:
yes, however it's frequently not a good idea. The stuff on line 15 will add cmake/ to the module path and allow including things from there.

You can also do what extra-cmake-modules does and maintain that stuff as a seperate project.

The issue with having a parent cmakelists file set stuff like visibility and whatnot is that it means the child isn't an independent project anymore, which can be OK, but is something to balance.

sergey-shandar replied:
IMHO, for Azure SDK we would like to have it because we may have about 100 of projects in Azure SDK for C. So, to be consistent, we would like to share code/scripts/configuration as much as possible between Azure SDK projects.

sergey-shandar replied:
@bsiegel could you share your experience on this subjects?

danewalton replied:
might we want to try and limit the shared configurations/code/scripts to things only absolutely necessary across all repos? say we have the c and cpp standards set in the top level since it is a requirement across all azure c repos. but any other config attempting to be put in the top level should have strong justification since it will affect all repos.

barcharcraz replied:
IMHO what we should have is something like extra-cmake-modules in the "core" repo.

So some core project (which we'll probably end up with anyway) will provide cmake macros/functions and will be a build-dependency of all/most other azure-c projects.

Projects could say something like
```cmake
if(NOT DEFINED AZ_CORE_MODULE_PATH)
    find_package(az_core REQUIRED CONFIG)
endif()
list(APPEND CMAKE_MODULE_PATH ${AZ_CORE_MODULE_PATH})
```

then do as they need.

@ras0219-msft what do you think about this? Is this a good way to go about it?

CaseyCarter commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+#include "operations.h"
+int32_t az_exampleshortname_add_two_numbers(int32_t a, int32_t b) {
+    return a + b;
+}
```
```suggestion
}

```

CaseyCarter commented on project_template/test/test_operations.cpp:
```diff
@@ -0,0 +1,9 @@
+
+TEST_CASE("operations test", "[operations]") {
+    REQUIRE(az_exampleshortname_add_two_numbers(1, 1) == 2);
+}
```
```suggestion
}

```

CaseyCarter commented on project_template/test/test_operations.cpp:
```diff
@@ -0,0 +1,9 @@
+#include <catch2/catch.hpp>
+#include <az/exampleshortname/operations.h>
+
+
```
Why two empty lines here?

CaseyCarter commented on project_template/test/main.cpp:
```diff
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier: MIT
+#define CATCH_CONFIG_MAIN
+
+#include <catch2/catch.hpp>
```
```suggestion
#include <catch2/catch.hpp>

```

sergey-shandar commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+cmake_minimum_required(VERSION 3.15)
+
+project(az_exampleshortname
+    LANGUAGES C CXX
```
Should we limit it to `C` only?

ras0219-msft replied:
If `CXX` is removed here, then when building tests you'll need to `enable_language(CXX)`[1]. I've not actually used this command and it looks like there are some complex restrictions on exactly how it must be used -- it is probably easiest to lift the `BUILD_TESTING` option above the project directive and conditionally enable `CXX` based on that setting.

[1] https://cmake.org/cmake/help/latest/command/enable_language.html

sergey-shandar commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+add_library(az_exampleshortname)
+add_library(az::exampleshortname ALIAS az_exampleshortname)
+include(GenerateExportHeader)
+generate_export_header(az_exampleshortname 
```
Could we avoid code generation on build?

joshfree commented on .vscode/snippets.code-snippets:
```diff
@@ -0,0 +1,26 @@
+	// Placeholders with the same ids are connected.
+	// Example:
+	// "Print to console": {
+	// 	"scope": "javascript,typescript",
```
can we change the example to C99

barcharcraz replied:
I just removed the example, it was autogenerated by vscode. Could also remove the snippet but imo it's handy

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+endif()
+
+#generate documentation
+if(az_exampleshortname_with_docs)
```
Should be `BUILD_DOCS` to be symmetric with `BUILD_TESTING` etc.

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+endif()
+
+# generate the install steps
+include(KDEInstallDirs)
```
This is not worth pulling in ECM. I would recommend removing this dependency and using `GNUInstallDirs` directly (built-in to cmake).

ras0219-msft commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,12 @@
+
+This azure sdk module has several external build-time requirements.
+
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
```
If the desire is to standardize on a single testing framework, Catch is less likely to be sufficient to cover all requirements (for example, death tests). Instead, I would recommend Google Test as the de-facto mature, complete, and maintained testing library. Additionally, Visual studio has explicit support for google test and does not have that support for catch.

If there isn't a desire to standardize on a single testing framework, then suggesting Catch2 here is less problematic.

barcharcraz replied:
yeah, I can switch it to gtest, but ultimately it's similar.

rafilho replied:
If we decide to standardize, then we need to have a deeper discussion. Our test framework should also be portable and small enough to run on some of our devices. Catch2 itself doesn't seem to be the best choice for a C library, since it does not support C officially. We have a homegrown test and mock frameworks which we have been using for 4+ years which could also be an option.

---
In reply to: [308971910](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r308971910) [](ancestors = 308971910)

ras0219-msft commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,12 @@
+This azure sdk module has several external build-time requirements.
+
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
+- **extra-cmake-modules**: https://github.com/KDE/extra-cmake-modules.git
```
It seems undesirable to require any external dependencies to perform the core library build if possible. I'd like to see this removed as a core requirement and only possibly used when building "extra" stuff (such as documentation).

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+    if(NOT TARGET Catch2::Catch2)
+        find_package(Catch2 REQUIRED CONFIG)
+    endif()
+    include(Catch)
```
Why is this required?

barcharcraz replied:
whoops. for catch_discover_tests, which makes ctest work and which I forgot to call.

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+cmake_minimum_required(VERSION 3.15)
```
This may be too high of a minimum version to meet the required audience.

For example, the (only?) available ubuntu version in Azure DevOps Pipelines' Hosted Pool is 18.04 LTS. This version of ubuntu only has CMake 3.10 available in the core repositories[1].

The VS 2015 hosted image has CMake 3.7.1[2].
The VS 2017 hosted image has CMake 3.14.5[3].
The VS 2019 hosted image has CMake 3.14.5[4].

[1] https://packages.ubuntu.com/search?keywords=cmake&searchon=names&suite=bionic&section=all
[2] https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2015-Server2012R2-Readme.md
[3] https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2017-Server2016-Readme.md
[4] https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2019-Server2019-Readme.md

sergey-shandar replied:
We can install required version in AzureDevOps, of course if we need higher than 3.12. Ubuntu 16 now has 3.12.

barcharcraz replied:
I think I'm OK with 3.14.5, but we want to decide if we want VERSION 3.14.5...3.15 or if we want to manually turn on CMP0091 when on 3.15 or above.

https://cmake.org/cmake/help/v3.15/policy/CMP0091.html#policy:CMP0091

ras0219-msft replied:
Also note: I was wrong about the version of ubuntu available -- looks like they only have 16.04 explicitly listed.

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+        ${CMAKE_CURRENT_SOURCE_DIR}/inc/az/exampleshortname
+        ${CMAKE_CURRENT_BINARY_DIR}/inc/az/exampleshortname)
+if(BUILD_TESTING)
+    include(CTest)
```
What more does this do than `enable_testing()` here? Why do we need that extra functionality?

barcharcraz replied:
I thought it was better to include it instead of enable_testing, I can change it though.

barcharcraz replied:
narrator: it was not. Robert is right again.

ras0219-msft commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+/** @file operations.h */
+#pragma once
+#include <stdint.h>
+#include <az/exampleshortname/export.h>
```
It seems worth checking this file in instead of generating it.

barcharcraz replied:
it may want to live in some core library. If we have specific export headers for each client library (for example to allow building some as static and some as dynamic) then it could make sense to have cmake generate the export headers.

ras0219-msft replied:
After chatting offline, this should be as simple as
```
// project_template/src/az/export.h
#if !defined(BUILDING_DLL)
#define AZ_EXPORT
#elif defined(_MSC_VER)
#define AZ_EXPORT __declspec(dllexport)
#else
#define AZ_EXPORT __attribute__((visibility("default")))
#endif
```
This file won't be installed; it will only be consumed in implementation files and any exported definitions should be marked with `AZ_EXPORT`. The public header files should all be completely un-annotated and should not include this file.

ras0219-msft commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include "operations.h"
```
If this is intending to include the public header, this should `#include <az/exampleshortname/operations.h>`.

barcharcraz replied:
the reason to allow this is that it lets private headers become public without changing their source files. I like this style, but we can say "just use full name"

ras0219-msft commented on .vscode/snippets.code-snippets:
```diff
@@ -0,0 +1,26 @@
+	"copyright and license header": {
+		"prefix": "copyright",
+		"body": [
+			"$LINE_COMMENT Copyright (c) Microsoft Corporation. All rights reserved.",
```
Does this do the right thing for C? (`/* */`)

CaseyCarter replied:
I don't know what this does for C, but I know that C has had `//` comments for 20 years.

ras0219-msft replied:
Right, C99 minimum version :) Thanks for the note!

ras0219-msft commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+    LANGUAGES C CXX
+    VERSION 0.1.0)
+
+option(az_exampleshortname_with_docs "Build documentation" ON)
```
Lowercase options feel "weird" to me (as someone who reads a lot of CMake files).

barcharcraz replied:
I like them. Only downside is #cmakedefine

anporumb commented on project_template/inc/az/exampleshortname.h:
```diff
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier: MIT
+
+/** @file exampleshortname.h Azure example library */
+#include <az/exampleshortname/operations.h>
```
>< [](start = 9, length = 1)

Chapter 6.10.2 Source file inclusion in C11 standard contains all the details about the difference between include< and include ".

In this case consider having include ".

barcharcraz replied:
in this case would you want #include "exampleshortname/operations.h"?

anporumb commented on project_template/inc/az/exampleshortname.h:
```diff
@@ -0,0 +1,5 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/** @file exampleshortname.h Azure example library */
```
>exampleshortname [](start = 10, length = 16)

Consider having include guards so that the header file is not included twice

barcharcraz replied:
this header doesn't include them because there should be nothing except includes of other code here. I'm not opposed to adding them (or `#pragma once`) but I don't feel it's necessary.

anporumb commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
+
+/** @file operations.h */
+#pragma once
```
not all compiler support #pragma once. However, all compilers will support 
#ifndef OPERATIONS_H
#define OPERATIONS_H..


#endif

barcharcraz replied:
pragma once is supported by all modern compilers except for Cray Classic and TASKING VX. If we want to support those compilers then it is very, very easy to switch from #pragma once to include guards using a little bit of regex.

Also `#pragma once` is much better in the case where you copied / moved / symlinked / somehow got a file from more than one place, since it will likely give you warnings / errors instead of silently compiling.

sergey-shandar replied:
`#pragma once` is awesome feature but I thought we should limit ourselves to C99.

mamokarz replied:
Did you try that on Arduino compilers, including the ESP8266 and ESP32? MCU manufactures didn't claim that their compilers are fully C99 and they may not have these kind of `pragma`.

barcharcraz replied:
Those seem to use a gcc based toolchain which would support this pragma

anporumb commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+
+/** @file operations.h */
+#pragma once
+#include <stdint.h>
```
Consider having #include <cstdint> when a C++ interface is provided

barcharcraz replied:
sorry I don't understand?

Perhaps this comment got pinned to the wrong line on github.

mamokarz replied:
Perhaps is because you should include `stdint.h` for C and `cstdint` for C++.

barcharcraz replied:
Well in this case we want `<stdint.h>` in C++ as well. <cstdint> is not guaranteed to define ::uint32_t, only std::uint32_t.

anporumb replied:
The lastest C++standard version that I have (C++14) states in "18.4.1 Header <cstdint> synopsis" the following:

"The header defines all functions, types, and macros the same as 7.18 in the C standard."

So based on the above I would reasonably expect that ::uint32_t/uint32_t (in C++-ese) type exists.

---
In reply to: [309482553](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309482553) [](ancestors = 309482553)

anporumb replied:
fun fact: C++98, C++03 are both mum about cstdint. It appears in C++11. So I guess it all depends on which version of C++ is available/targeted.

---
In reply to: [310337767](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310337767) [](ancestors = 310337767,309482553)

dcristoloveanu replied:
From what I see Charlie is right. Unfortunately including cstdint is not enough, one would have to have a using namespace std to have the same code compile.
So just including stdint.h in C++ too is actually a better option I think.

It would have been nice that the C++ standard would state that cstdint should define ::int32_t etc. Because these being in std forces one of the 3:
- Using stdint.h which defeats the purpose of cstdint
- Adding std:: everywhere for C++, which defeats again the purpose of cstdint
- Have using namespace std all over the place which can introduce other potential collisions

Sooooo, yeah. The lesser evil is #include <stdint.h> I guess. Now we have some code bases to update ...

---
In reply to: [310337826](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310337826) [](ancestors = 310337826,310337767,309482553)

barcharcraz replied:
yep. <cmath> is somewhat of a special case since it defines additional overloads. This is also unfortunate but if you're c++ code and want to use those overloads then you need cmath, and not math.h.

fwiw if you're a C library, providing a C interface, it's reasonable to assume your compiler supports C (and its standard library).

dcristoloveanu replied:
C++ (at least 14) say stdint.h is provided (and the others). Annex D, D.1 which is normative.
So it should be safe to include stdint.h all the time (in C and C++). 

It is unfortunate that cstdint does not help in this case for C++ as that would be ideal. There are cases where a translation unit must be compiled with the C++ compiler and still needs to include headers from the C lib. Happened several times when porting IoT SDK for C that such interactions were needed. From a nicety and consistency point of view would have been nice to be able to use cstdint in those cases, but ... it is what it is.

---
In reply to: [310861226](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310861226) [](ancestors = 310861226)

anporumb commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include "operations.h"
```
>"operations.h" [](start = 9, length = 14)

Is this intended to be used like this?

It seems that the project folder structure is crafted to provide a namespace-like construct.

Is this intended to actually be #include "az/exampleshortname/operations.h"?

anporumb commented on project_template/test/test_operations.cpp:
```diff
@@ -0,0 +1,8 @@
+#include <catch2/catch.hpp>
+#include <az/exampleshortname/operations.h>
+
+TEST_CASE("operations test", "[operations]") {
```
>(" [](start = 9, length = 2)

Consider testing C code with a C framework.

This seems to be a C++ framework.

barcharcraz replied:
yes it is. The testing frameworks, and their integration into tools (like cmake, visual studio, qt creator, eclipse, azure pipelines, github, jenkins, vim, emacs, etc is much better than any C testing framework I know about (closest would be mutest and gtest (gnome-test, not google test).

We don't consider it unreasonably to require a C++ compiler to build the tests.

danewalton commented on project_template/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/external/**/
+/build*
```
should the .gitignore have the cmake/ directory as well?

ras0219-msft replied:
typically `cmake/` is used to contain any extra cmake modules that need to be checked in (i.e. source code), so you wouldn't want to add that to gitignore.

barcharcraz replied:
actually it probably shouldn't include /build-aux

danewalton commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+extern "C" {
+#endif
+/** @brief add two numbers together
+ *  @param[in] a
```
>a [](start = 15, length = 1)

consider putting a description of the parameters (like type and purpose of the parameter)
eg: The first `uint32_t` to add.

rafilho replied:
We should strive for having enough annotation for generating our documentation with doxygen. @danewalton, can you please paste a link for our generated doxygen documentation here as a illustration?

---
In reply to: [309387205](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309387205) [](ancestors = 309387205)

danewalton replied:
[Example header](https://github.com/Azure/azure-ulib-c/blob/master/inc/ustream.h)
[Documentation for that file and repo](https://azure.github.io/azure-ulib-c/ustream_8h.html)

---
In reply to: [309420922](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309420922) [](ancestors = 309420922,309387205)

barcharcraz replied:
I would argue that this is pretty well covered by the @return value.

danewalton replied:
possibly for this example but consider a function that returns some library return value like `FOO_RESULT`. You can use the doxygen tag `@return` to describe the type of the return and then `@retval` to describe the different values that could be returned. [Example](https://github.com/Azure/azure-ulib-c/blob/6240cb4b7a74562dcd08374b49ac605708bc023a/inc/ustream.h#L87)

barcharcraz replied:
yes I think that's a good idea.

Do you want any action here ,for this specific function?

danewalton replied:
nitpicky but formatting it like [this](https://github.com/Azure/azure-ulib-c/blob/master/inc/ustream.h) does makes it nicely readable and gives a good amount of information. so things like a newline between the different tags, detailing the type of the parameters and the return parameter. we also have things tab aligned between the different components like for the params, it would be `@param[in] <TAB> a <TAB> The first <tt>uint32_t</tt> to add`

danewalton commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+/** @brief add two numbers together
+ *  @param[in] a
+ *  @param[in] b
+ *  @return @p a + @p b
```
>@p [](start = 12, length = 2)

Similar to the parameters, consider putting the returned type here. Someone reading the documentation will want to know what the type is of the return value.

barcharcraz replied:
isn't that already covered by the declaration of the function? The return type will be included in the docs automatically.

I think doxygen may actually even include the `int` return type for functions declared in K&R style (default int), but we won't be allowing that.

danewalton replied:
it does not. it will put the function with the input/output, function declaration etc in the top level but you need to manually write the documentation for the different parts of it. you can see how we do it [here](https://github.com/Azure/azure-ulib-c/blob/master/inc/ustream.h) with [these](https://azure.github.io/azure-ulib-c/ustream_8h.html) results.

barcharcraz replied:
OK I like using @retval for describing functions with special return values. But here there are no special return values.

danewalton replied:
understood and agree. but i still think the standard we set should be that we lay out explicitly the type of the return. so here it might be `The @int32_t with the result of the addition` or something like that.

anporumb commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+    INSTALL_DESTINATION ${AZ_INSTALL_CMAKEPACKAGEDIR}/${PROJECT_NAME})
+install(TARGETS az_exampleshortname EXPORT ${PROJECT_NAME}-targets)
+install(DIRECTORY inc DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+install(EXPORT ${PROJECT_NAME}-targets DESTINATION ${AZ_INSTALL_CMAKEPACKAGEDIR}/${PROJECT_NAME} NAMESPACE az::)
```
>AZ_INSTALL_CMAKEPACKAGEDIR [](start = 53, length = 26)

Consider generally having BUILDING one action and PACKAGING the result of the build a separate action.

ras0219-msft replied:
These commands result in exactly that:
```sh
cmake .. # Configure step, detects compilers and so forth
make # Just builds the binaries
make install # Extracts the built binaries into a package
```

barcharcraz replied:
we could have a option to turn off install target generation, but there's not really a good reason to do that.

rafilho replied:
+1

---
In reply to: [309394596](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309394596) [](ancestors = 309394596)

anporumb commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,11 @@
+
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
+
+These may be installed using a package manager or embedded in the external/ folder.
```
>external [](start = 66, length = 8)

Consider sub-repo'ing (submodules) all the possible external dependencies. This comes with the great advantage of pinning the used submodule to a specific SHA.

If left to the package manager or user or some other means that will break the build dependency as far as versioning goes.

ras0219-msft replied:
The critical scenarios are:
1. A user should be able to clone, build, and consume *just this library* easily with pinned, tested versions.
2. A user should be able to integrate this library into a larger project which already has copies of `curl`, `libuv`, etc. This is critical to enable the user to resolve security issues by updating dependencies.

Solving (2) implies compatibility with package managers. One way to solve (1) is to submodule all external dependencies. An alternative way to solve (1) is to use an embeddable package manager, such as hunter or vcpkg, though it will obviously be critical to not compromise the "SHA pinning" aspect required by point (1).

barcharcraz replied:
I actually have support for submodule dependencies. The reason I don't is that I wanted this project to be a subfolder in azure-sdk-for-c, and didn't want to add submodules to the parent repository. I could add a gitmodules file to make them easier to add.

Also @ras0219-msft imo using submodules has the additional advantage that it makes it easier to reduce / fix bugs in your dependencies and submit those fixes upstream.

Another reason I didn't add the submodules is that if the library is building as part of a large build (say the whole azure-sdk-for-c project) then each individual library shouldn't be vendoring anything.

anporumb commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,11 @@
+
+This Azure SDK module has several external build-time requirements.
+
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
```
>tests [](start = 18, length = 5)

catch2 doesn't seem to come with a mocking framework. What's the story around that requirement?

barcharcraz replied:
the story is "we haven't decided on a test framework, I wanted a test example, the buildsystem integration is almost identical for all/most test frameworks, and I had a project using catch2 open on my other monitor"

rafilho commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include "operations.h"
+int32_t az_exampleshortname_add_two_numbers(int32_t a, int32_t b) {
```
>{ [](start = 66, length = 1)

Are we enforcing the curly bracket style? If yes, IMO, we should go with Allman style (curly braces in new line for each block) in our samples. This is too Java-ish.

barcharcraz replied:
I have added a clang-format file to the root (and reformatted source files). It's based on the LLVM style.

danewalton replied:
seconded

---
In reply to: [309417312](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309417312) [](ancestors = 309417312)

rafilho replied:
Any motivations/arguments for choosing this specific style? 

---
In reply to: [309447254](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309447254) [](ancestors = 309447254)

barcharcraz replied:
LLVM is what we use in the STL, and we've discussed this in meetings with the azure-native team and they seem to like it. I don't really care as long as everyone agrees and there's a clang-format file.

https://clang.llvm.org/docs/ClangFormatStyleOptions.html has documentations on the various styles and clang-format options.

danewalton replied:
there are parts of LLVM that seem to clash with several different code bases we have. 
First LLVM uses camel case in which case it clashes with this file. Our team prefers snake case like you use here since that also aligns with threadx and their general practices. We also use Allman style braces as does threadx.

---
In reply to: [309470098](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309470098) [](ancestors = 309470098)

barcharcraz replied:
The llvm style (and clang format in general) will never, ever, ever rename a type. 

We (the vclibs and the azure-native teams) are not trying to import and follow the llvm style guide, we were just OK with using their formatting style as a base.

rafilho commented on project_template/test/test_operations.cpp:
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include <catch2/catch.hpp>
```
>catch2 [](start = 10, length = 6)

Is the plan to enforce the test harness/mocking platform? Catch2 is a C++ test runner. It does not even officially support C:

Catch2 is a multi-paradigm test framework for C++. which also supports Objective-C (and maybe C). It is primarily distributed as a single header file, although certain extensions may require additional headers. ([https://github.com/catchorg/Catch2#whats-the-catch](https://github.com/catchorg/Catch2#whats-the-catch)).

In IoT, we have a custom test runner for C (released as open source in GitHub) which is very slim (meaning that we can port to a constrained device and run tests there): [https://github.com/Azure/azure-ctest](https://github.com/Azure/azure-ctest). We also have a mocking framework ([https://github.com/Azure/umock-c](https://github.com/Azure/umock-c)).

dcristoloveanu commented on project_template/inc/az/exampleshortname/operations.h:
```diff
@@ -0,0 +1,20 @@
+ *  @param[in] b
+ *  @return @p a + @p b
+ */
+AZ_EXAMPLESHORTNAME_EXPORT int32_t az_exampleshortname_add_two_numbers(int32_t a, int32_t b);
```
>AZ_EXAMPLESHORTNAME_EXPORT [](start = 0, length = 26)

Question: what is this for? Is it for abstracting declspecs?
In the past in the IotHub SDK for C team when we had to have exports for dynamic libraries we did that exclusively through .def files in order to not introduce extra artifacts in the code.
And we did not need other declspecs TBH as we shied away from most platform specific things.

barcharcraz replied:
this is very likely to change due to feedback from Robert.

On windows, this will add __declspec(dllimport) which makes calling the function a little faster.

ras0219-msft replied:
As additional information that may be relevant if we want to maintain .def files for windows, it looks like there is an available equivalent for other platforms:

```
gcc -Wl,--version-script=symbols.map
```

https://stackoverflow.com/questions/792195/gnu-linker-alternative-to-version-script-to-list-exported-symbols-at-the-comm

I would encourage this direction if strong ABI stability is a goal.

dcristoloveanu commented on project_template/BUILDING.md:
```diff
@@ -0,0 +1,11 @@
+
+This Azure SDK module has several external build-time requirements.
+
+- **Catch2** (for tests): https://github.com/catchorg/Catch2.git
```
>* (for tests): https://github.com/catchorg/Catch2.git [](start = 11, length = 53)

Regarding tools/frameworks that have been used in the past:
The Azure IoTHub SDK team uses the following:

- azure-ctest ([https://github.com/Azure/azure-ctest](https://github.com/Azure/azure-ctest)). This is a very simple and portable test runner for C. At the time of its writing we did consider CppUnitTest as the test runner but the fact that it was C++ and it required small changes for porting even when building with gcc was a no-go for portability.
- azure-c-testrunnerswitcher ([https://github.com/Azure/azure-c-testrunnerswitcher](https://github.com/Azure/azure-c-testrunnerswitcher)). This is a switcher between ctest and CppUnitTest. The reason it exists is because sometimes it is desirable to run, debug analyze tests under VS.
- umock-c ([https://github.com/Azure/umock-c](https://github.com/Azure/umock-c)). This is a C mocking library. It offers recording of calls, comparing expected and actual calls, negative tests, handling various types (charptr, stdint, bool, custom types), allows chaining of modifiers on expectations, etc.

All of the tests in IoTHub C SDK are written with these tools and they have served us well for the past 4-5 years.

Let me know if you find these valuable - I can ramble about this waaaay more :-).

dcristoloveanu commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+cmake_minimum_required(VERSION 3.15)
+
+project(az_exampleshortname
+    LANGUAGES C CXX
```
> C CXX [](start = 13, length = 6)

This being C SDK this should be only C.

dcristoloveanu commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include "operations.h"
+int32_t az_exampleshortname_add_two_numbers(int32_t a, int32_t b) {
```
>int32_t [](start = 0, length = 7)

Must include <stdint.h>.
Each source file should handle its own includes otherwise there are hidden dependecies that can be easily broken.

dcristoloveanu commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+cmake_minimum_required(VERSION 3.15)
```
>3.15 [](start = 31, length = 4)

Are there any features really needed from 3.15?
The only reason I am saying this is that in the IoTHub SDK case there were several platforms that had lower versions of cmake (one of them had 2.8 at some point). Consider using the lowest version according to the needed features from cmake.

danewalton replied:
We should hopefully (bare minimum) be using something above 3.0. The more recent versions of cmake make use of targets and properties rather than passing along variables.

---
In reply to: [309437127](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309437127) [](ancestors = 309437127)

barcharcraz replied:
The current most likely to standardize on is 3.6.

dcristoloveanu commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_CXX_STANDARD 17)
```
Usually the enforced standard is the minimal version required for the code. C99 seems fine, just wondering why do we require C++ 17? This would exclude cases where customers want to use a lower version - assuming our code does not really require C++17 of course.

barcharcraz replied:
that's not what CMAKE_CXX_STANDARD does unless you also set CMAKE_CXX_STANDARD_REQUIRED (same for C, but with C instead of CXX).

dcristoloveanu replied:
Got it. Looked closer at these - learn something new everyday.

---
In reply to: [309450713](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309450713) [](ancestors = 309450713)

barcharcraz commented on project_template/src/operations.c:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MIT
+#include "operations.h"
+int32_t az_exampleshortname_add_two_numbers(int32_t a, int32_t b) {
+    return a + b;
```
This would complicate the example, the intent of this function is to provide a symbol that actually ends up in the library to confirm shared libraries work (and to give something for the tests to call).

While I agree that we may want to handle overflow I don't think think it's necessary for the sample project.

rafilho commented on project_template/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+
+set_property(DIRECTORY PROPERTY COMPILE_OPTIONS
+    $<$<COMPILE_LANG_AND_ID:C,MSVC>:/W4 /wd4201>
+    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Xclang -Wall -Wextra -Wno-microsoft-anon-tag -Wdocumentation>)
```
what is going to be our strategy for cross compilation?

barcharcraz replied:
cmake toolchain files.

msalehmsft commented:
@jebrando 
@ewertons 
@jspaith 
@tameraw 
@mamokarz 
@massand 
@jasmineymlo 
@danewalton 
@hihigupt

barcharcraz commented:
Given that there doesn't seem to be consensus on testing I have removed the example test setup and replaced it with a usage example.

I've also changed the operation in question to be defined for all inputs and documented as such.

All generated files have been removed from the build (aside from the doxyfile, and the cmake configuration file) and .def files or linker version scripts are now used to export symbols when building shared libraries. The default visibility presets have been changed back to public from hidden, because they override linker scripts.

barcharcraz commented:
That’s not what [cstdint.syn] (now 17.4.1) means.

What “The header defines all types and macros the same as the C standard library header <stdint.h>” means is that
Any types defined (::std::whatever) will be the same as ::whatever in <stdint.h>, that is, if you include both
Cstdint and stdint.h ::std::whatever is the same as ::whatever.



From: anporumb <notifications@github.com>
Sent: Friday, August 2, 2019 8:55 PM
To: Azure/azure-sdk-for-c <azure-sdk-for-c@noreply.github.com>
Cc: Charlie Barto <Charles.Barto@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [Azure/azure-sdk-for-c] Project template (#2)


@anporumb commented on this pull request.

________________________________

In project_template/inc/az/exampleshortname/operations.h<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-c%2Fpull%2F2%23discussion_r310337767&data=02%7C01%7CCharles.Barto%40microsoft.com%7C617ff0e4b504464b6d8e08d717c652f4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637004012870845539&sdata=ZCjPNWOlagjTYzC%2FZjOHj0WsX41BhC18m7Q6mBf8zHg%3D&reserved=0>:

> @@ -0,0 +1,20 @@

+// Copyright (c) Microsoft Corporation. All rights reserved.

+// SPDX-License-Identifier: MIT

+

+/** @file operations.h */

+#pragma once

+#include <stdint.h>

The lastest C++standard version that I have (C++14) states in "18.4.1 Header synopsis" the following:

"The header defines all functions, types, and macros the same as 7.18 in the C standard."

So based on the above I would reasonably expect that ::uint32_t/uint32_t (in C++-ese) type exists.

________________________________

In reply to: 309482553<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-c%2Fpull%2F2%23discussion_r309482553&data=02%7C01%7CCharles.Barto%40microsoft.com%7C617ff0e4b504464b6d8e08d717c652f4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637004012870855532&sdata=aAeBk%2F%2FCUdawI7cbowdaIOpgZRyy77amY%2Fd4BwdaJ8o%3D&reserved=0> [](ancestors = 309482553)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-c%2Fpull%2F2%3Femail_source%3Dnotifications%26email_token%3DAAJ6PPOSBFSXWQCI4PNHXODQCT6QLA5CNFSM4IIAL6B2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCAPKUUA%23discussion_r310337767&data=02%7C01%7CCharles.Barto%40microsoft.com%7C617ff0e4b504464b6d8e08d717c652f4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637004012870855532&sdata=RyxVZMJpxl%2FkwaNaPZicr0utBHPVzpKVzDopCzXNGHA%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAJ6PPKXVLQB3CJFCAQKHHDQCT6QLANCNFSM4IIAL6BQ&data=02%7C01%7CCharles.Barto%40microsoft.com%7C617ff0e4b504464b6d8e08d717c652f4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637004012870865528&sdata=F96hJW%2BkRsrYPSmzogz91AnNe15eVF%2BUhrvekSrBJNs%3D&reserved=0>.

bsiegel commented:
Only able to comment from an engsys standpoint - looks good to me, but we should locate this project under the conventional directory `sdk/<servicename>/<libraryname>`. For a template like this we could just use `sdk/template/template`.
